### ISSUE TITLE
Fitting DIG to image

### ISSUE THREAD

**Issue by kristijanbartol** [liren2515/DIG#4]
Hi Ren,

Thank you very much for preparing the public source code. I only recently read your paper and checked the code. It seems that there should a part of the code where you implement Eq. 15 for fitting the model to the image, but I find it hard to verify which part of the code it is and how exactly to use it for fitting. 

I would like to use this fitting to evaluate your model on other images for quantitative but also qualitative comparison. Can you please help me with specifying the part of the code? I believe it has to do with the following function: https://github.com/liren2515/DIG/blob/main/models/model_sdf.py#L184, but I can't easily figure out where the Eq. 15 and how to use it.

Thank you once again!

**Comment by liren2515** [comment]
Hello,

Thanks for your interest in our work. Actually, the codes for fitting are not included. But you can customize your own code with our models and a differentiable renderer (like pytorch3D) for this purpose.  

The fitting is quite straightforward:
1. initialize a latent code z
2. reconstruct the mesh by our sdf model:
   ```
   cloth_mesh = reconstruct(z, model_sdf)
   gar_verts, gar_faces = cloth_mesh.vertices, cloth_mesh.faces 
   ```
3. bring back the differentiability following [meshsdf](https://github.com/cvlab-epfl/MeshSDF):
   `new_vertices = get_mesh_sdf(gar_verts, gar_faces, z, model_sdf)`
4. deform `new_vertices` with the `deform` function in [here](https://github.com/liren2515/DIG/blob/main/utils/deform.py) to get the deformed vertices `vertices_deformed`:
   `verts_gar_deformed, _, _, _= deform(...)`
5. render the mask for the mesh (a concatenation of body and the cloth).
6. compute the loss and do backprop.

The following is an incomplete example script for fitting customized for pytorch3D (here we only optimize the latent code for the cloth, and you need to set up your own renderer to use it):
```
from pytorch3d.structures import Meshes
from pytorch3d.renderer import TexturesVertex

def update_z_shirt(pose, beta, z_style, joints_2D_gt, images_gt, smpl_tfs, tfs_c_inv, shapedirs, tfs_weighted_zero, embedder, model_lbs, model_lbs_delta, model_blend_weight, renderer_textured, iters=100):

    lr = 1e-2
    z_style.requires_grad = True 
    pose.requires_grad = False 
    beta.requires_grad = False 
    optimizer = torch.optim.Adam([{'params': z_style, 'lr': lr}])

    with torch.no_grad():
        smpl_verts_pred, joints_pred, _, _, smpl_tfs = smpl_server.forward_verts(betas=beta,
                                                transl=transl,
                                                body_pose=pose[:, 3:],
                                                global_orient=pose[:, :3],
                                                return_verts=True,
                                                return_full_pose=True,
                                                v_template=smpl_server.v_template, rectify_root=False)
        smpl_verts_pred = smpl_verts_pred.squeeze()
        smpl_tfs = smpl_tfs.squeeze()
        
        smpl_verts_pred.requires_grad = False 
        smpl_tfs.requires_grad = False 

    loss_min = 1e10
    z_style_best = z_style.clone().detach()
    for step in range(iters):

        with torch.no_grad():
            cloth_mesh = reconstruct(z_style, model_G, just_vf=False, resolution=256)
            gar_verts, gar_faces = cloth_mesh.vertices, cloth_mesh.faces 

        verts_zero = torch.zeros(len(smpl_verts_pred)+len(gar_verts), 3)
        faces = torch.cat((smpl_faces.cpu(), torch.LongTensor(gar_faces) + len(smpl_verts_pred)))
        # assign labels for body and cloth
        smpl_rgb = torch.zeros(len(smpl_verts_pred), 3)
        smpl_rgb[:,0] += 255
        gar_rgb = torch.zeros(len(gar_verts), 3)
        gar_rgb[:,1] += 255
        verts_rgb = torch.cat((smpl_rgb, gar_rgb))[None]
        textures = TexturesVertex(verts_features=verts_rgb.to(device))

        mesh = Meshes(
            verts=[verts_zero.to(device)],   
            faces=[faces.to(device)],
            textures=textures
        )

        new_vertices = get_mesh_sdf(gar_verts, gar_faces, z_style, model_G)

        verts_gar_deformed, cloth_mesh, delta_theta, delta_beta = deform(new_vertices, gar_faces, smpl_tfs, tfs_c_inv, pose, beta, shapedirs, tfs_weighted_zero, embedder, model_lbs, model_lbs_delta, model_blend_weight)

        verts_deformed = torch.cat((smpl_verts_pred, verts_gar_deformed), dim=0)
        # this is just for coordinate matching since SMPL and py3d have different definition for the world coordinate.
        signs = torch.ones_like(verts_deformed).cuda()
        signs[:,:2] *= -1
        verts_deformed *= signs

        new_src_mesh = mesh.offset_verts(verts_deformed)
        images_predicted = renderer_textured(new_src_mesh)
        images_pred = images_predicted[0, :, :, :3]/255
        
        gar_silh = images_pred[:, :, 1]
        gar_gt = images_gt[:, :, 1]
        intersection_g = (gar_silh*gar_gt).sum()
        union_g = gar_silh.sum() + gar_gt.sum() - intersection_g
        loss_texture = (1 - intersection_g/union_g)*224#10
        loss_reg = z_style.norm()/10

        line = 'step: %3d, loss_texture: %0.4f, loss_reg: %0.5f'%(step, loss_texture.item(), loss_reg.item())
        print(line)

        loss = loss_texture + loss_reg 

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
    return z_style_best
```

and this is the function to bring back the differentiability:
```
def get_mesh_sdf(verts, faces, style, model_G):

    ### 2: evaluate analytical normals
    verts_torch = torch.from_numpy(verts).float().cuda()
    verts_torch.requires_grad = True
    num_points = len(verts_torch)
    x_cloth_points = style.unsqueeze(0).repeat(num_points, 1)
    pred_sdf_verts = model_G(verts_torch, x_cloth_points, num_points)
    pred_sdf_verts.sum().backward(retain_graph=True)

    normals = verts_torch.grad
    normals = normals / (torch.norm(normals, dim=1, keepdim=True) + 0.0001)

    ### 3: new verts= old vert + sdf (=0) * normals, to bring bck differentiability
    new_verts = verts_torch.detach() - pred_sdf_verts * normals

    return new_verts
```

**Comment by kristijanbartol** [comment]
Hi Ren,

Thank you very much for your detailed answer. I will try it out in the following few days, but the information you provided should be sufficient to make sense of it, so I will close the issue. :)

**Comment by NagabhushanSN24** [comment]
What is `renderer_textured`? Can you give an example of what can be used for it? Should it be a differentiable module?

**Comment by liren2515** [comment]
> What is `renderer_textured`? Can you give an example of what can be used for it? Should it be a differentiable module?

You can use the renderer from `pytorch3d`:

```
from pytorch3d.structures import Meshes
from pytorch3d.renderer import (
    PerspectiveCameras,
    RasterizationSettings, 
    MeshRenderer, 
    MeshRasterizer
)

class SimpleShader(nn.Module):
    def __init__(self, device="cpu", blend_params=None, hard_mode=False):
        super().__init__()
        self.blend_params = blend_params if blend_params is not None else BlendParams()
        self.hard_mode = hard_mode

    def forward(self, fragments, meshes, **kwargs) -> torch.Tensor:
        blend_params = kwargs.get("blend_params", self.blend_params)
        #print(blend_params.background_color)
        texels = meshes.sample_textures(fragments)
        if self.hard_mode:
            images = hard_rgb_blend(texels, fragments, blend_params)
        else:
            images = softmax_rgb_blend(texels, fragments, blend_params)
        return images  # (N, H, W, 3) RGBA image

cameras = PerspectiveCameras(device=device, focal_length=focal, in_ndc=False, image_size=[render_res], principal_point=princpt)

sigma = 1e-5
raster_settings_soft = RasterizationSettings(
    image_size=render_res, 
    blur_radius=np.log(1. / 1e-4 - 1.)*sigma, 
    faces_per_pixel=20, 
    max_faces_per_bin=500000,
    perspective_correct=False
)

renderer_textured = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings_soft
    ),
    shader=SimpleShader(device=device, hard_mode=False)
)
```

**Comment by NagabhushanSN24** [comment]
Thank you!
